DevTools DI Visualization Features

AleksanderBodurri opened the pull request:
See individual commits.

Note that the majority of the LOC additions come from mock data in unit tests.

yekanchi commented:
any screenshot?

dgp1130 reviewed (changes requested):
Thanks for putting this all together! Definitely a lot of d3 complexity here. I'm only about halfway through right now. Most of it are nitpicks / random code quality suggestions. Feel free to push back on any of these, I don't think any are blocking this PR.

The biggest question I have is what the impact of the `@angular/core` fix is on the versioning of the release. Does this mean we need the latest patch version to do any DI debugging? Is it fixing a specific bug in a specific situation or is this more load-bearing?

I would love to get a review from someone who understands more about DI. I can't provide a whole lot of feedback outside of general observations. Maybe Andrew Kushnir could take a look whenever you think it's ready?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.scss:
```diff
@@ -0,0 +1,415 @@
+        font-weight: 300;
+      }
+    }
+}
```
Nit: Indentation. We're inconsistently using 2 or 4 spaces.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.scss:
```diff
@@ -0,0 +1,415 @@
+}
+
+:host {
+  ::ng-deep {
```
Question: Is `:host ::ng-deep` necessary, or can we just do `::ng-deep`?

AleksanderBodurri replied:
This is safer to avoid leaking styles outside of this component

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.scss:
```diff
@@ -0,0 +1,415 @@
+  }
+}
+
+:host {
```
Suggestion: This is a big file and selectors like `:host` and `::ng-deep` seem to repeat a lot. I suggest either grouping all the common selectors together or sectioning them into related styles. For example, `.deps`, `.no-deps`, `:host-context(.dark-theme) { .deps { } }` can be reasonably grouped together, even if other dark theme stuff exists elsewhere.

dgp1130 reviewed (changes requested):
Ok, finally got through the rest of it. Again, it's mostly pretty minor nitpicks and quality suggestions, nothing very critical. A couple comments are probably related to the fact that the PR is still in draft state and not fully complete.

dgp1130 commented:
Actually the biggest thing is probably the lack of tests. Testing the d3 stuff might be difficult and arguably not worth it, but we could certainly test a lot of the injection tree transformations and component behavior outside of rendering.

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/component-tree.ts:
```diff
@@ -103,6 +147,169 @@ export const getDirectiveMetadata = (dir: any): DirectiveMetadata => {
+ */
+export function getNodeInjectorTNode(nodeInjector) {
+  return (nodeInjector as any)._tNode;
+}
```
Definitely agree with this. There are a few places where code was duplicated to prevent public API exposure. I think it'd be a good idea to minimize this as much as possible but did not want to block this initial feature

AleksanderBodurri replied:
resolved by making a new debug API in #51900

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/dependency-injection/injector-tree-visualizer.ts:
```diff
@@ -0,0 +1,230 @@
+      private _containerElement: HTMLElement, private _graphElement: HTMLElement,
+      private _orientation: 'horizontal'|'vertical' = 'horizontal',
+      private _nodeSize: [number, number] = [70, 200], private _nodeSeperation?: number,
+      private _childFn?: any) {
```
Good catch, forgot to remove this while refactoring.

jessicajaniuk reviewed (approved):
reviewed-for: fw-core

dgp1130 reviewed (approved):
Everything looks promising, so I'm very excited to this go in! I'll approve now just so this isn't blocked while I'm OOO, but take your time on the implementation. Definitely looking forward to seeing this land!

jessicajaniuk commented:
@AleksanderBodurri Looks like this still has some CI issues.

AleksanderBodurri commented:
Will do a bit more cleanup here related to @dgp1130 's feedback before we merge

AleksanderBodurri committed:
fix(core): framework debug APIs getDependenciesForTokenInInjector and getInjectorMetadata

Previously getDependenciesForTokenInInjector was unable to determine which node on a view serviced a specific injection. Now it is able to filter out those injections that did not come from the specific node for the NodeInjector passed into it.

Previously getInjectorMetadata was incorrectly looking up DOM elements for some directives (for example NgForOf) where an LContainer was created. Now the LContainer case is handled, and the non LContainer case uses `getFirstNativeNode` to more accurately get the element we want.

AleksanderBodurri committed:
feat(devtools): Implement initial DI debugging features in devtools

This commit introduces 2 new features into DevTools.

Directive level dependency inspection: Users can now view which dependencies their directives have injected in the property viewer tab. This view displays not only the dependency but also the resolution path that was used to service the injection.

Injector graph inspection: Users can now view a visualization of the element and environment hierarchies in their application. These trees are displayed separately but on the same page in the Injector Tree tab. User can click on individual injectors to view a list of all the providers configured in that injector, as well as highlight the resolution path from that injector to the root (with the corresponding environment injector connection highlighted as well).

AleksanderBodurri committed:
test(devtools): create unit tests for injector tree transformation functions

Creates set of unit tests for each function in the data transformation pipeline that enables injector metadata to be visualized as d3 graphs.

AndrewKushnir commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -413,9 +448,12 @@ export function getInjectorMetadata(injector: Injector):
+      return {type: 'element', source: unwrapRNode(lView[tNode.index]) as RElement};
+    }
+
+    return {type: 'element', source: getFirstNativeNode(lView, tNode) as RElement};
```
Could you please add a comment into the code to explain why we do this and not just taking `lView[tNode.index]` node?

AleksanderBodurri replied:
I thought `getFirstNativeNode` would be good to use here because it covers a lot of cases around going from LView and Tnode to a DOM element. Looking again though it seems that `getNativeByTNode` fits the idea of an NodeInjector `source` better. Was able to remove the `isLContainer` check this way as well.

`getNativeByTNode` is well named. Should I still add a comment here?

AndrewKushnir replied:
> getNativeByTNode is well named. Should I still add a comment here?

Agreed, we don't need extra comments in this case, the code looks clear.

AndrewKushnir commented on packages/core/src/render3/util/injector_discovery_utils.ts:
```diff
@@ -413,9 +448,12 @@ export function getInjectorMetadata(injector: Injector):
 
-    return {type: 'element', source: lView[tNode.index][HOST]};
+    if (isLContainer(lView[tNode.index])) {
+      return {type: 'element', source: unwrapRNode(lView[tNode.index]) as RElement};
```
I'd propose a more explicit call to extract `[HOST]`:
```
const lContainer = lView[tNode.index];
return {type: 'element', source: unwrapRNode(lContainer[HOST]) as RElement};
```
The `unwrapRNode` has the logic to read `HOST` inside, but it looks a bit confusing here which RNode we are trying to retrieve.

AndrewKushnir commented on packages/core/src/render3/di.ts:
```diff
@@ -743,9 +743,8 @@ export function getNodeInjectorLView(nodeInjector: NodeInjector): LView {
-    TElementContainerNode|null {
-  return (nodeInjector as any)._tNode as TElementNode | TContainerNode | TElementContainerNode |
-      null;
+    TElementContainerNode {
+  return (nodeInjector as any)._tNode as TElementNode | TContainerNode | TElementContainerNode;
```
Curious why do we change the type here.

AndrewKushnir commented:
FYI, just pushed a rebase to rerun CI, the PR was based on a commit that had CI failure.

AndrewKushnir commented:
[Presubmit](http://test/OCL:571959751:BASE:571959787:1696869092350:b4bb75ab).

AndrewKushnir commented:
@AleksanderBodurri it looks like there is a legit error in core acceptance tests, see [this CI job output](https://github.com/angular/angular/actions/runs/6459224486/job/17534573373?pr=51719). Could you please take a look when you get a chance?

AndrewKushnir commented:
Update: this PR needs cleanup: there is a legit case were `TNode` might be `null`, but we don't account for this in the code (this triggers CI to fail). @AleksanderBodurri has additional info on this.

AndrewKushnir commented:
[Presubmit](http://test/OCL:572316152:BASE:572316014:1696962550457:aee1d768).

AndrewKushnir commented:
**Caretaker note**: presubmit and CI is "green", this PR is ready for merge.

haraldreingruber-dedalus commented:
Hello. Thanks a lot for this contribution!

I would really like to test this on our project, as it might help us a lot for our refactoring.
I think the devtools in the chrome webstore wasn't updated in several months.
Can you help me figuring out the easiest way to run it locally?
I guess I have to update our project to one of the latest angular releases because of the backend part. Is the transpiled/compiled code of the devtools extension somewhere I can download and install locally?

Thanks a lot!

JeanMeche commented:
The sources for the extensions are available on the repo. Here are the build instructions :  https://github.com/angular/angular/blob/main/devtools/README.md

dgp1130 commented:
@haraldreingruber-dedalus just FYI, this will be published alongside Angular v17 very shortly: https://angular.io/guide/releases#support-policy-and-schedule

Also this particular feature requires your app to be run on v17, so you'll need to upgrade first. You can test out `17.0.0-rc` if you want to try the prerelease.

We'd definitely like to hear any feedback you have from trying it, I'm just letting you know what you're in for if you try to use this early.

haraldreingruber-dedalus commented:
@AleksanderBodurri @dgp1130
I was finally able to update our app to Angular v17. 
The injector-tree tab of the DevTools is working, and looks super nice so far 😎

Should I expect to see all entities of the DI container?
Could it be that `@Injectable({providedIn: 'root'})` objects are missing, or do I misunderstand how it works?

Or maybe many important entities are missing because our project is split into several libraries?
